Source/NSGridView.m

```diff
@@ -0,0 +1,1280 @@
+{
+  return [[self alloc] initWithViews: rows];
+}
+
+- (NSInteger) numberOfRows
+{
+  return [_rows count];
+}
+
+- (NSInteger) numberOfColumns
+{
+  return [_columns count];
+}
+
+- (NSGridRow *) rowAtIndex: (NSInteger)index
+{
+  return [_rows objectAtIndex: index];
+}
+
+- (NSInteger) indexOfRow: (NSGridRow *)row
+{
+  return [_rows indexOfObject: row];
+}
+
+- (NSGridColumn *) columnAtIndex: (NSInteger)index
+{
+  return [_columns objectAtIndex: index];
+}
+
+- (NSInteger) indexOfColumn: (NSGridColumn *)column
+{
+  return [_columns indexOfObject: column];
+}
+
+- (NSGridCell *) cellAtColumnIndex: (NSInteger)columnIndex rowIndex: (NSInteger)rowIndex
+{
+  NSUInteger idx = rowIndex * [self numberOfColumns] + columnIndex;
+  return [_cells objectAtIndex: idx];
+}
+
+- (NSGridCell *) cellForView: (NSView*)view
+{
+  NSGridCell *result = nil;
+
+  FOR_IN(NSGridCell*, c, _cells)
+    {
+      NSView *v = [c contentView];
+      if (v == view)
+        {
+          result = c;
+          break;
+        }
+    }
+  END_FOR_IN(_cells);
+
+  return result;
+}
+
+- (NSGridRow *) addRowWithViews: (NSArray *)views
+{
+  return [self insertRowAtIndex: [self numberOfRows]
+                      withViews: views];
+}
+
+- (NSRect) _findProtypeView
+{
+  NSRect f = NSMakeRect(0,0,10,10);
+
+  // Find a cell to base the size off of.
+  if ([_cells count] > 0)
+    {
+      FOR_IN(NSGridCell*, acell, _cells)
+        {
+          NSView *av = [acell contentView];
+          if (av)
+            {
+              f = [av frame];
+              break;
+            }
+        }
+      END_FOR_IN(_cells);
+    }
+  
+  return f;
+}
+
+- (NSGridRow *) insertRowAtIndex: (NSInteger)index withViews: (NSArray *)views
+{
+  NSGridRow *gr = [[NSGridRow alloc] init];
+
+  // Insert the row and release...
+  [_rows insertObject: gr atIndex: index];
+  [gr setGridView: self];
+  RELEASE(gr);
+
+  NSRect f = [self _findProtypeView];
+  NSUInteger i = 0;
+  NSUInteger pos = index * [self numberOfColumns];
+  FOR_IN(NSView*, v, views)
+    {
+      NSGridCell *c = [[NSGridCell alloc] init];
+      NSGridColumn *col = [_columns objectAtIndex: i];
+      
+      [col setGridView: self];
+      [c _setOwningRow: gr];
+      [c _setOwningColumn: col];
+      [v setFrame: f];
+      [c setContentView: v];
+      RELEASE(v);
+      [_cells insertObject: c
+                   atIndex: pos + i];
+      RELEASE(c);
+      i++;
+    }
+  END_FOR_IN(views);
+
+  // Insert remaineder of cells for views not present..
+  NSUInteger r = [self numberOfColumns] - i;
+  NSUInteger idx = 0;
+  pos += i;
+  for(idx = 0; idx < r; idx++)
+    {
+      NSGridCell *c = [[NSGridCell alloc] init];
+      [_cells insertObject: c atIndex: pos + idx];
+    }
+  
+  // Refresh...
+  [self _refreshCells];
+  return gr;
+}
+
+- (void) moveRowAtIndex: (NSInteger)fromIndex toIndex: (NSInteger)toIndex
+{
+  NSArray *views = [self _viewsForRowAtIndex: fromIndex];
```

Comment:
This is a possible way to implement this method, but that way we loose all the information that was set on a specific cell.

Reply:
Fixed.  The methods now deal with the cells directly.